pdns/lua-recursor4.cc

```diff
@@ -1033,3 +1034,150 @@ void pdns_ffi_param_add_meta_single_int64_kv(pdns_ffi_param_t* ref, const char*
 {
   ref->params.meta[std::string(key)].intVal.insert(val);
 }
+
+struct pdns_postresolve_ffi_handle
+{
+public:
+  pdns_postresolve_ffi_handle(RecursorLua4::PostResolveFFIHandle& h) :
+    handle(h)
+  {
+  }
+  RecursorLua4::PostResolveFFIHandle& handle;
+  auto insert(std::string&& str)
+  {
+    const auto it = pool.insert(std::move(str)).first;
+    return it;
+  }
+
+private:
+  std::unordered_set<std::string> pool;
+};
+
+bool RecursorLua4::postresolve_ffi(RecursorLua4::PostResolveFFIHandle& h) const
+{
+  if (d_postresolve_ffi) {
+    pdns_postresolve_ffi_handle_t handle(h);
+
+    auto ret = d_postresolve_ffi(&handle);
+    return ret;
+  }
+  return false;
+}
+
+const char* pdns_postresolve_ffi_handle_get_qname(pdns_postresolve_ffi_handle_t* ref)
+{
+  auto str = ref->insert(ref->handle.d_dq.qname.toStringNoDot());
+  return str->c_str();
+}
+
+uint16_t pdns_postresolve_ffi_handle_get_qtype(const pdns_postresolve_ffi_handle_t* ref)
+{
+  return ref->handle.d_dq.qtype;
+}
+
+uint16_t pdns_postresolve_ffi_handle_get_rcode(const pdns_postresolve_ffi_handle_t* ref)
+{
+  return ref->handle.d_dq.rcode;
+}
+
+pdns_policy_kind_t pdns_postresolve_ffi_handle_get_appliedpolicy_kind(const pdns_postresolve_ffi_handle_t* ref)
+{
+  return static_cast<pdns_policy_kind_t>(ref->handle.d_dq.appliedPolicy->d_kind);
+}
+
+void pdns_postresolve_ffi_handle_set_appliedpolicy_kind(pdns_postresolve_ffi_handle_t* ref, pdns_policy_kind_t kind)
+{
+  ref->handle.d_dq.appliedPolicy->d_kind = static_cast<DNSFilterEngine::PolicyKind>(kind);
+}
+
+bool pdns_postresolve_ffi_handle_get_record(pdns_postresolve_ffi_handle_t* ref, unsigned int i, pdns_ffi_record_t* record, bool raw)
+{
+  if (i >= ref->handle.d_dq.currentRecords->size()) {
+    return false;
+  }
+  try {
+    DNSRecord& r = ref->handle.d_dq.currentRecords->at(i);
+    record->name = ref->insert(r.d_name.toStringNoDot())->c_str();
```

Comment:
See other comment. This is a field that we would never need when checking response IPs.

Reply:
We likely could provide  the raw name by default, and have a separate method to get the human readable form.

Reply:
I think that would make sense. I suppose providing the raw name would be almost free?

Reply:
> I think that would make sense. I suppose providing the raw name would be almost free?

Indeed, sonce we anot creating a tremp code there's also no need to store the string into the string pool. 

I changed the code to always return a name_len and fill a raw name or a human readable name depending on the raw flag, so both name and content are now either raw or human readable.

Reply:
Can the raw name have back references in this context? If so, the raw value may not be useful without a separate function to convert them.

Reply:
And if we are going that way, we could also just omit the name and have a separate function to fetch the name with an index parameter.

Reply:
No back refs here, we're dealing with objects at this stage. The recursor will create the answer packet later.